ez-launch/sources/ez_launch.move

```diff
@@ -0,0 +1,462 @@
+
+        // (jill) check against whitelist
+
+        execute_coin_payments(user, collection_details_obj);
+
+        let token_index = if (collection_details_obj.random_mint) {
+            timestamp::now_seconds() % length
+        } else {
+            length - 1
+        };
+        let token = *vector::borrow(&available_tokens, token_index);
+        let user_address = signer::address_of(user);
+        object::transfer(&object_signer, token, user_address);
+        vector::pop_back(&mut available_tokens);
+
+        if (collection_details_obj.is_soulbound) {
+            token_components::freeze_transfer(&object_signer, token);
+        };
+
+        token
+    }
+
+    fun assert_owner<T: key>(owner: address, object: Object<T>) {
+        assert!(object::owner(object) == owner, error::permission_denied(ENOT_OWNER));
+    }
+
+    fun authorized_collection_signer(creator: &signer, collection: Object<Collection>): signer acquires EZLaunchRefs {
+        let refs = authorized_borrow(creator, collection);
+        object::generate_signer_for_extending(&refs.extend_ref)
+    }
+
+    fun collection_signer(collection: Object<Collection>): signer acquires EZLaunchRefs {
+        let refs = borrow(collection);
+        object::generate_signer_for_extending(&refs.extend_ref)
+    }
+
+    fun add_mint_fee(
+        mint_fee: &mut Option<u64>,
+        creator_addr: address,
+        collection_details: &mut CollectionDetails,
+    ) {
+        if (option::is_some(mint_fee)) {
+            let mint_fee_category = b"Mint Fee";
+            let coin_payment = coin_payment::create<AptosCoin>(
+                option::extract(mint_fee),
+                creator_addr,
+                string::utf8(mint_fee_category),
+            );
+            vector::push_back(&mut collection_details.coin_payments, coin_payment);
+        };
+    }
+
+    fun execute_coin_payments(
+        user: &signer,
+        collection_details: &mut CollectionDetails,
+    ) {
+        vector::for_each_ref(&collection_details.coin_payments, |coin_payment| {
+            let coin_payment: &CoinPayment<AptosCoin> = coin_payment;
+            coin_payment::execute(user, coin_payment);
+        });
+    }
+
+    fun royalty(
+        royalty_numerator: &mut Option<u64>, 
+        royalty_denominator: &mut Option<u64>, 
+        creator_addr: address
+    ): Option<Royalty> {
+        assert!(option::is_some(royalty_numerator) == option::is_some(royalty_denominator), error::invalid_argument(EROYALTY_CONFIGURATION_INVALID));
+        if (option::is_some(royalty_numerator) && option::is_some(royalty_denominator) && option::extract(royalty_numerator) != 0 && option::extract(royalty_denominator) != 0) {
+            option::some(royalty::create(option::extract(royalty_numerator), option::extract(royalty_denominator), creator_addr))
+        } else {
+            option::none()
+        }
+    }
+
+    fun configure_collection_and_token_properties(
```

Comment:
man this is still so ugly lol. @JohnChangUK

Reply:
Ah I see. But the whole point of default values was so that the user does not have to provide these at the time of creating a collection. I think it should be in separate function calls, to split out the creation of the collection vs the properties.

Better yet, if the user provides these details, we should not be checking if it is true or false, but just setting them all directly, by calling each setter. This means that the user has already decided to set the collection properties and once configured, should be immutable :o